extensions/amp-list/0.1/amp-list.js

```diff
@@ -1122,12 +1122,16 @@ export class AmpList extends AMP.BaseElement {
 
   /**
    * Measure and lock height before performing given mutate fn.
-   * Applicable for layout=container.
+   * Applicable for layout=container without infinite scrolling.
    * @private
    * @param {!Function} mutate
    * @return {!Promise}
    */
   lockHeightAndMutate_(mutate) {
+    devAssert(
+      !this.loadMoreEnabled_,
+      'amp-list[layout=container] does not support infinite scrolling with [load-more].'
```

Comment:
Is there a way to make this a validator error or something so publishers (esp. those that already tried to do this and had it break) will know?

Reply:
I know that some attributes can require others, but I'm not aware of syntax for attributes that disallow others. If it's possible we could insert a validation rule to describe "if `load-more` is present, don't allow `layout` to have value `container`". This wouldn't break any existing pages because layout=container was only enabled as an attribute value last week.

It gets tricky because there are two other ways that amp-list can _change_ its layout _to_ "container". One way is through the `is-layout-container` attribute, and we may not want this to be mutually exclusive because it could break existing valid pages. Another way is through the `changeToLayoutContainer` action which I believe cannot be targeted via validations.

We could print a warning that, once the amp-list becomes `layout=container` with one of the above methods, checks if `load-more` is enabled. And update documentation to more clearly reflect that these are mutually exclusive features for the time being. 

cc/ @honeybadgerdontcare

Reply:
@caroqliu Why are they mutually exclusive features? I've used `changeToLayoutContainer` when the user clicks the "load more" button. I would think at that point, there's been a user interaction, so an immediate resize of the amp-list would be OK.

Reply:
@chasefinch Thanks for weighing in, could you please share a code sample? It would be great to have a better grasp on how people rely on amp-list features to work together to avoid fixes like were needed from this PR.

My understanding is that when you use the `load more` feature on `amp-list`, it handles the resizing for you. [This sample](https://amp.dev/documentation/examples/interactivity-dynamic-content/show_more_button/) describes two ways to implement a "Show more" button on your page, without needing `changeToLayoutContainer` -- does using that action provide more to enhance the experience?

Reply:
Good point that amp-list can handle the resizing for new items on its own. In my case, one reason for wanting layout `container` is that child elements still may have expanding elements that require a resize of the `amp-list`.

But then in practice, swapping to layout container also alleviates the pain of some outstanding bugs in amp-list:

- [[amp-list] Glitch using [src] and load more together](https://github.com/ampproject/amphtml/issues/27778)
- [Make \<amp-list\> resizeable.](https://github.com/ampproject/amphtml/issues/25662) / [amp-list doesn't reclaim space after expanding browser](https://github.com/ampproject/amphtml/issues/27251)
- [amp-list: overflow show-more not working with load-more attribute](https://github.com/ampproject/amphtml/issues/23420)

And finally, `changeToLayoutContainer` lets the list expand to fit any further new items even if it would cause a screen jump - which I realize is by design, but in my case it's nice that the users always get the results that they asked for, regardless of screen position. In general, I think that amp-list should get some kind of concession on the no-jump rule, since it represents dynamic external content that is (in the case of a manual load-more button) explicitly requested.

This is a simplified version of the code I'm using:
```
<amp-list src="/source/"
  id="resultsList"
  width="auto"
  height=800
  layout="fixed-height"
  load-more="manual"
  noloading
>
  <div placeholder>
    <p>Loading...</p>
  </div>

  <template type="amp-mustache">
    {{ '#items' }}
      <details>
        <summary>{{ title }}</summary>

        A block of content that’s initially hidden, but would change the necessary amp-list height when shown.
      </details>
    {{ '/items' }}
  </template>

  <amp-list-load-more load-more-button>
    <button on="resultsList.changeToLayoutContainer">Load More</button>
  </amp-list-load-more>

</amp-list>
```

Reply:
It seems to be that the issue is in the naming of the features `is-layout-container` and `changeToLayoutContainer` conflating the approach to achieve the desired consequence with the intention behind the action. In this case the intention could be "force resize to fit contents" and the approach is "accomplish this by setting `layout` to container", with the implicit bridge being that an element with `layout` container assumes the size of its contents. 

By introducing an unconventional application of `layout=container` that is restricted to avoid content jumping that is not initiated by user interaction, we break the underlying assumption critical for achieving the desired action of forcing resize to fit contents.

Seeing as we can't retroactively change the names of these actions, allowing layout=container in its restrictive capacity is ... totally confusing from a language and devX perspective. 

Perhaps instead of allowing layout=container from the get go we could introduce the feature under a different guise, i.e. an attribute such as `"enable-managed-resizing"`? cc @choumx

Reply:
Sorry, what's the actual delta here? AFAICT, the things missing from `layout=container` are (1) allowing the amp-list to shrink in height vs. only grow and (2) a mechanism to "unlock" the amp-list height when resize fails. 

Is there anything else preventing `layout=container` from superseding `is-layout-container`/`changeToLayoutContainer`?

Reply:
I'm not sure what you mean by "superseding" as it applies to live usage. Enabling `layout=container` with the 2 amendments you described would serve to make `is-layout-container`/`changeToLayoutContainer` unnecessary going forward, but still impacts that pages use it today. 

The `amp-list` internals doesn't account for `layout=container` because it was initially set versus because it was toggled by `is-layout-container`/`changeToLayoutContainer`. 

If we allow publishers to set `layout=container` initially, we need to account for the existing API with one of these stances:

- You can't use `is-layout-container`/`changeToLayoutContainer`, but instead you can use a new action `resizeToContents` that internally essentially mimics `is-layout-container`/`changeToLayoutContainer`. This could also conceptually replace `is-layout-container`/`changeToLayoutContainer` but deprecating its predecessors altogether seems unlikely.
- You can use `is-layout-container`/`changeToLayoutContainer` but it's actually a misnomer since it would be used only to mean "force unlock height"

Wouldn't both of these options be confusing?

Reply:
Yea I think we might agree. I meant `layout=container` could replace them in amp-list V2 which would avoid this API complexity.